### PR TITLE
BUG: Fix PeriodIndex +/- TimedeltaIndex

### LINKED ISSUE #22998: BUG: PeriodIndex + offset incorrect
Troubleshooting based on `tests.indexes.period.test_tools.TestPeriodIndexConversion.test_to_timestamp_pi_combined`

```
idx = PeriodIndex(start='2011', periods=2, freq='1D1H', name='idx')

>>> idx + 1
PeriodIndex(['2011-01-02 01:00', '2011-01-03 02:00'], dtype='period[25H]', name=u'idx', freq='25H')

>>> idx + idx.freq
PeriodIndex(['2011-01-27 01:00', '2011-01-28 02:00'], dtype='period[25H]', name=u'idx', freq='25H')
```

`idx + idx.freq` should match `idx + 1`.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #22998
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/util/testing.py [hunk 1]
```diff
@@ -805,6 +805,7 @@ def assert_index_equal(left, right, exact='equiv', check_names=True,
         Specify object name being compared, internally used to show appropriate
         assertion message
     """
+    __tracebackhide__ = True
```
Comment (jbrockmendel) [inline pandas/util/testing.py]: See #22643

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -631,11 +631,14 @@ def __add__(self, other):
...
             elif is_integer_dtype(other):
                 result = self._addsub_int_array(other, operator.add)
-            elif is_float_dtype(other) or is_period_dtype(other):
+            elif is_float_dtype(other):
                 # Explicitly catch invalid dtypes
                 raise TypeError("cannot add {dtype}-dtype to {cls}"
                                 .format(dtype=other.dtype,
                                         cls=type(self).__name__))
+            elif is_period_dtype(other):
+                # PeriodIndex + TimedeltaIndex _sometimes_ is valid
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py]: can you clarify?

**Review by jreback — CHANGES_REQUESTED** [review]
add a whatsnew note

**Comment by jbrockmendel** [comment]
Updated with a whatsnew and more tests.  Doing this right required more a refactor than I expected, but in the end it touches only arithmetic code and doesn't fiddle with indexing code (like the previous attempt did)

File: pandas/core/arrays/period.py [hunk 3]
```diff
@@ -355,24 +356,54 @@ def _add_offset(self, other):
...
+        return DatetimeLikeArrayMixin._add_delta_td(self, delta)
+
+    def _add_delta_tdi(self, other):
+        assert isinstance(self.freq, Tick)  # checked by calling function
+
+        delta = self._check_timedeltalike_freq_compat(other)
+        return self._addsub_int_array(delta, operator.add)
 
     def _add_delta(self, other):
-        ordinal_delta = self._maybe_convert_timedelta(other)
```
Comment (jbrockmendel) [inline pandas/core/arrays/period.py]: Getting the _maybe_convert_timedelta call out of the arithmetic ops makes everything simpler

File: pandas/core/arrays/period.py [hunk 4]
```diff
@@ -355,24 +356,54 @@ def _add_offset(self, other):
...
+        #  i8 view or _shallow_copy
+        if isinstance(other, (Tick, timedelta, np.timedelta64)):
+            new_values = self._add_delta_td(other)
+            return self._shallow_copy(new_values)
+        elif is_timedelta64_dtype(other):
+            # ndarray[timedelta64] or TimedeltaArray/index
+            new_values = self._add_delta_tdi(other)
+            return self._shallow_copy(new_values)
+        else:  # pragma: no cover
+            raise TypeError(type(other).__name__)
```
Comment (jbrockmendel) [inline pandas/core/arrays/period.py]: Looking at this and some of the other methods in the datetime/timedelta/period array mixins, I think some de-duplication can be done on the next pass

**Comment by jbrockmendel** [comment]
gentle ping

File: pandas/core/arrays/period.py [hunk 5]
```diff
@@ -454,6 +480,38 @@ def _maybe_convert_timedelta(self, other):
         raise IncompatibleFrequency(msg.format(cls=type(self).__name__,
                                                freqstr=self.freqstr))
 
+    def _check_timedeltalike_freq_compat(self, other):
+        assert isinstance(self.freq, Tick)  # checked by calling function
```
Comment (jreback) [inline pandas/core/arrays/period.py]: can u add a doc string here

File: pandas/core/arrays/period.py [hunk 6]
```diff
@@ -454,6 +480,38 @@ def _maybe_convert_timedelta(self, other):
...
+        own_offset = frequencies.to_offset(self.freq.rule_code)
+        base_nanos = delta_to_nanoseconds(own_offset)
+
+        if isinstance(other, (timedelta, np.timedelta64, Tick)):
+            nanos = delta_to_nanoseconds(other)
+
+        elif isinstance(other, np.ndarray):
```
Comment (jreback) [inline pandas/core/arrays/period.py]: can u comments these cases a bit

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. just add a couple of comments. merge when green

**Comment by jbrockmendel** [comment]
Ping

**Pull request merged**